### PR TITLE
roles: hosted_engine_setup: Add after_add_host hook

### PR DESCRIPTION
**Pull request by didib**
Allow running custom playbooks after trying to add the host to the
engine.

Bug-Url: https://bugzilla.redhat.com/1893385

### CODE REVIEW

**Comment by didib** [comment]
Didn't verify yet.

**Comment by arachmani** [comment]
@didib, The code looks good to me, but I didn't understand why we need to add another place for running hooks if the workaround from https://bugzilla.redhat.com/show_bug.cgi?id=1893385#c0 works.
In case we need this PR, I think It's best to add some explanation also in the HE role's [README](https://github.com/oVirt/ovirt-ansible-collection/blob/master/roles/hosted_engine_setup/README.md) file, something similar to https://github.com/oVirt/ovirt-ansible-collection/tree/master/roles/hosted_engine_setup#make-changes-in-the-engine-vm-during-the-deployment

**Comment by didib** [comment]
> @didib, The code looks good to me, but I didn't understand why we need to add another place for running hooks if the workaround from https://bugzilla.redhat.com/show_bug.cgi?id=1893385#c0 works.

The workaround is not a full solution - after you apply it, you still have to eventually setup the networks as needed and mark them required. And during this, they are not required. With current patch, the user has an option to automatically handle failures _after_ trying to add the host, presumably (I didn't try yet) by assigning relevant NICs to the required networks and trying to activate, therefore never marking them as non-required.

I intend to also add another "wait for lockfile" (similar to he_pause_host, but perhaps mandatory, we'll see), to also allow manually fixing, for people that prefer that.

> In case we need this PR, I think It's best to add some explanation also in the HE role's [README](https://github.com/oVirt/ovirt-ansible-collection/blob/master/roles/hosted_engine_setup/README.md) file, something similar to https://github.com/oVirt/ovirt-ansible-collection/tree/master/roles/hosted_engine_setup#make-changes-in-the-engine-vm-during-the-deployment

Sure.

**Comment by didib** [comment]
I don't like github. It seems like the only way for me to mark this PR as not-ready-for-merge is to change it to a draft. In gerrit I can '-1'. Here, if I try to add a comment and mark "Request Changes", a popup says "Pull request authors can't request changes on their own pull request". Searching for this text in google finds very few other people that care. Github simply managed to dumb-down everyone to not expect more, it seems.

**Comment by arachmani** [comment]
> I don't like github. It seems like the only way for me to mark this PR as not-ready-for-merge is to change it to a draft. In gerrit I can '-1'. Here, if I try to add a comment and mark "Request Changes", a popup says "Pull request authors can't request changes on their own pull request". Searching for this text in google finds very few other people that care. Github simply managed to dumb-down everyone to not expect more, it seems.

Maybe you can just add `[WIP]` in the subject line.
And when the PR is ready just remove it.
You can do that easily from the UI.

**Comment by mwperina** [comment]
I've created a WIP label for such patches, let's agree that not maintainer will merge a patch with WIP label. Is it OK for you?

**Comment by didib** [comment]
OK for me.

**Comment by arachmani** [comment]
OK for me.

**Comment by didib** [comment]
It seems that only people with write access to the repo can assign labels:

https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/applying-labels-to-issues-and-pull-requests

(And no, the answer isn't "OK, so we'll give you write access". IMO).

**Comment by mwperina** [comment]
Martin, is there some easy way how to have a bot watching for certain string in comments and setting labels accordingly? Something similar to ManageIQ bot https://github.com/ManageIQ/miq_bot where they can use below comment to add a label

`@miq-bot add_labels label1`

@mnecas

**Comment by didib** [comment]
ci test please

**Comment by didib** [comment]
Verified both manual (let it fail adding, prompt, add via web ui and rm lock file) and automatic (using the example hook).

File: roles/hosted_engine_setup/README.md [hunk 1]
```diff
@@ -356,6 +362,10 @@ In order to proceed with the deployment, before deleting the lock-file, make sur
 
 Both of the lock-file path and the engine's URL will be presented during the role execution.
 
+**On Failure**
+
+If "Add Host" failed and left the host in status "non_operational", by default the deployment will be paused, similarly to "Manual" above, so that the user can try to fix the host to get it to "up" state, before removing the lock file and continuing. If you want the process to fail instead of pausing, set he_pause_after_failed_add_host to false.
```
Comment (arachmani) [inline roles/hosted_engine_setup/README.md:367]: Missig backticks for `he_pause_after_failed_add_host`

File: roles/hosted_engine_setup/defaults/main.yml [hunk 2]
```diff
@@ -49,6 +49,11 @@ he_force_ip4: false
 he_force_ip6: false
 
 he_pause_host: false
+
+# Pause execution similarly to setting he_pause_host to true, if "add host"
+# failed with status "non_operational". Set to false in your vars file if
+#  you want deploy to fail immediately instead of the default "true".
```
Comment (arachmani) [inline roles/hosted_engine_setup/defaults/main.yml:55]: I don't think we need to explain it here as we already have the explanation in README

File: roles/hosted_engine_setup/examples/required_networks_fix.yml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+# situation via the engine web admin UI.
+#
+# If you already know that beforehand, and want to automate restoration,
+# you can copy current file, edit as needed, and place it in:
+# /usr/share/ansible/collections/ansible_collections/ovirt/ovirt/roles/hosted_engine_setup/hooks/after_add_host/
+# File name should end with '.yml'.
+#
+# For more details, see (also):
+# https://docs.ansible.com/ansible/latest/collections/ovirt/ovirt/ovirt_host_network_module.html
+# https://docs.ansible.com/ansible/latest/collections/ovirt/ovirt/ovirt_host_module.html
```
Comment (arachmani) [inline roles/hosted_engine_setup/examples/required_networks_fix.yml:19]: The latest documentation is for version 2.10 and we are using 2.9.
I didn't check if there is any difference between them.

File: roles/hosted_engine_setup/examples/required_networks_fix.yml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    host_result_up_check.ovirt_hosts[0].status == 'non_operational'
+    )
+  retries: 120
+  delay: 10
+  ignore_errors: true
+
+- name: Handle non_operational myhost
+  block:
+  - name: Attach interface eth0 on host myhost to network net1
+    ovirt.ovirt.ovirt_host_network:
```
Comment (arachmani) [inline roles/hosted_engine_setup/examples/required_networks_fix.yml:42]: Can't we just use `ovirt_host_network:`?

File: roles/hosted_engine_setup/examples/required_networks_fix.yml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+      auth: "{{ ovirt_auth }}"
+      name: myhost
+      interface: eth0
+      networks:
+        - name: net1
+  - name: Activate host myhost
+    ovirt.ovirt.ovirt_host:
```
Comment (arachmani) [inline roles/hosted_engine_setup/examples/required_networks_fix.yml:49]: Ditto

File: roles/hosted_engine_setup/tasks/bootstrap_local_vm/05_add_host.yml [hunk 6]
```diff
@@ -156,13 +161,77 @@
       msg: >-
         Host is not up, please check logs, perhaps also on the engine machine
     when: host_result_up_check is failed
+
```
Comment (arachmani) [inline roles/hosted_engine_setup/tasks/bootstrap_local_vm/05_add_host.yml:164]: Remove redundant newline.

**Comment by marchukov** [comment]
ci please test

File: roles/hosted_engine_setup/README.md (hunk 1 above)
Comment (didib) [reply roles/hosted_engine_setup/README.md:367]: Done

File: roles/hosted_engine_setup/defaults/main.yml (hunk 2 above)
Comment (didib) [reply roles/hosted_engine_setup/defaults/main.yml:55]: ok, done

File: roles/hosted_engine_setup/examples/required_networks_fix.yml (hunk 3 above)
Comment (didib) [reply roles/hosted_engine_setup/examples/required_networks_fix.yml:19]: I generally prefer to link to latest.

That said, these modules are now maintained right here, in ovirt-ansible-collection, so it should continue working when we make changes.

I did wonder what ansible are going to do about documentation when splitting stuff out to collections, not sure.

File: roles/hosted_engine_setup/examples/required_networks_fix.yml (hunk 4 above)
Comment (didib) [reply roles/hosted_engine_setup/examples/required_networks_fix.yml:42]: I guess we can, but ansible recommends to use full name, and this is an example...

File: roles/hosted_engine_setup/examples/required_networks_fix.yml (hunk 5 above)
Comment (didib) [reply roles/hosted_engine_setup/examples/required_networks_fix.yml:49]: same

File: roles/hosted_engine_setup/tasks/bootstrap_local_vm/05_add_host.yml (hunk 6 above)
Comment (didib) [reply roles/hosted_engine_setup/tasks/bootstrap_local_vm/05_add_host.yml:164]: done

**Review by arachmani — APPROVED** [review]
LGTM

**Comment by didib** [comment]
Anything else? I think it's ready for merge. I don't mind squashing and/or splitting to separate PRs and opening bugs for some things, if you want (e.g. fixing error/warning collection).

**Pull request merged**